library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java

```diff
@@ -2535,4 +2505,76 @@ public int hashCode() {
 
   }
 
+  /** Represents how well an text track matches the selection {@link Parameters}. */
+  protected static final class TextTrackScore implements Comparable<TextTrackScore> {
+
+    private final boolean isWithinRendererCapabilities;
+    private final boolean isDefault;
+    private final boolean isForced;
+    private final int preferredLanguageScore;
+    private final int selectedAudioLanguageScore;
+    private final boolean trackHasNoLanguage;
+    private final boolean selectUndeterminedTextLanguage;
+    private final boolean stringDefinesNoLang;
+
+    public TextTrackScore(
+        Format format,
+        Parameters parameters,
+        int trackFormatSupport,
+        @Nullable String selectedAudioLanguage) {
+      isWithinRendererCapabilities = isSupported(trackFormatSupport, false);
+      int maskedSelectionFlags =
+          format.selectionFlags & ~parameters.disabledTextTrackSelectionFlags;
+      isDefault = (format.selectionFlags & C.SELECTION_FLAG_DEFAULT) != 0;
+      isForced = (maskedSelectionFlags & C.SELECTION_FLAG_FORCED) != 0;
+      preferredLanguageScore = getFormatLanguageScore(format, parameters.preferredTextLanguage);
+      selectedAudioLanguageScore = getFormatLanguageScore(format, selectedAudioLanguage);
+      trackHasNoLanguage = formatHasNoLanguage(format);
+      selectUndeterminedTextLanguage = parameters.selectUndeterminedTextLanguage;
+      stringDefinesNoLang = stringDefinesNoLanguage(selectedAudioLanguage);
```

Comment:
The comparison below which use a formatLanguageScore and the undetermined language is quite long. I think you could simplify this a lot by including it all into one "boolean hasLanguageMatch = preferredLanguageScore > 0 || (parameters.selectUndeterminedTextLanguage && trackHasNoLanguage)" and then just compare the boolean.

The same is true for the selectedAudioLanguage. You just need a "boolean hasSelectedAudioLanguageMatch = selectedAudioLanguageScore || (trackHasNoLanguage && stringDefinesNoLanguage(selectedAudioLanguage))"